### DISCUSSION TITLE
Generating python info from embedded modules

### DISCUSSION THREAD

**Discussion by asmaloney** [pybind/pybind11#5385] [Q&A]
I have a C++ application and I am embedding the interpreter as [outlined in the docs](https://pybind11.readthedocs.io/en/stable/advanced/embedding.html).

I have a module defined using `PYBIND11_EMBEDDED_MODULE`.

1. how do I extract whatever info I need so that VS Code "knows" about my module?
2. how do I run sphinx on this to generate docs for my module?

Everything I've found so far assumes that a python module is being generated - not embedded.

Thanks for any insights! (I'm a C++ guy, not a Python guy, so it's possible I'm misunderstanding some terminology.)

**Comment by asmaloney** [comment]
Does anyone have any ideas?

I have an embedded module that I want my users to be able to use to script parts of my application and it would be great to give them sphinx docs & let them work properly with an IDE.

I still haven't found solutions.

Thanks!

**Reply by AdrianDeWinter** [reply]
I've never worked with embedded modules, but have you taken a look at [pybind11-stubgen](https://github.com/sizmailov/pybind11-stubgen) ?
Once you have the .pyi files, IDEs and other tools should just happily work with those.

If you do go that way, this is useful: https://github.com/sizmailov/pybind11-stubgen/issues/227#issuecomment-2303532507

**Reply by asmaloney** [reply]
Thanks Adrian.

I did discover `.pyi` files along the way!

I had looked at `pybind11-stubgen`, but as far as I can tell there's no way to use it with embedded modules. It's expecting a module as an argument on the command line and there's no way to point at them. If I point at my executable I just get `ModuleNotFoundError` which makes sense 😄

**Reply by AdrianDeWinter** [reply]
Mhh, yeah, I read here that pybind11-stubgen just needs to be able to import the module: https://github.com/sizmailov/pybind11-stubgen/issues/227#issuecomment-2303532507
Checking against some other (built-in) libraries, running something like `pybind11-stubgen __future__`, it generates stubs just fine.

Thus, rather than passing your executable to stubgen, I would suggest using your embedded executable to run pybind11-stubgen (may need to find the right env var to set), and passing the importable name of your embedded module to it

**Reply by asmaloney** [reply]
I had tried something like that. It was a bit of a pain because I don't allow installation of any packages so I had to try to do that part manually.

In the end, it tried to call my application on the command line as if it were Python, which fork-bombed my computer.  Kinda funny watching dozens of my application icons bouncing in the dock, trying furiously to kill them, but ultimately not really useful 😆 

Quite possible I didn't have everything set up properly since it's kind of an inception-like setup at that point.
